File: dotnet/src/SemanticKernel/SemanticFunctions/PromptTemplate.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.SemanticKernel.Orchestration;
using Microsoft.SemanticKernel.TemplateEngine;

namespace Microsoft.SemanticKernel.SemanticFunctions;

/// <summary>
/// Prompt template.
/// </summary>
public sealed class PromptTemplate : IPromptTemplate
{
    private readonly string _template;
    private readonly IPromptTemplateEngine _templateEngine;

    // ReSharper disable once NotAccessedField.Local
    private readonly PromptTemplateConfig _promptConfig;

    /// <summary>
    /// Constructor for PromptTemplate.
    /// </summary>
    /// <param name="template">Template.</param>
    /// <param name="promptTemplateConfig">Prompt template configuration.</param>
    /// <param name="kernel">Kernel in which template is to take effect.</param>
    public PromptTemplate(string template, PromptTemplateConfig promptTemplateConfig, IKernel kernel)
        : this(template, promptTemplateConfig, kernel.PromptTemplateEngine)
    {
    }

    /// <summary>
    /// Constructor for PromptTemplate.
    /// </summary>
    /// <param name="template">Template.</param>
    /// <param name="promptTemplateConfig">Prompt template configuration.</param>
    /// <param name="promptTemplateEngine">Prompt template engine.</param>
    public PromptTemplate(
        string template,
        PromptTemplateConfig promptTemplateConfig,
        IPromptTemplateEngine promptTemplateEngine)
    {
        this._template = template;
        this._templateEngine = promptTemplateEngine;
        this._promptConfig = promptTemplateConfig;
    }

    /// <summary>
    /// Get the list of parameters used by the function, using JSON settings and template variables.
    /// TODO: consider caching results - though cache invalidation will add extra complexity
    /// </summary>
    /// <returns>List of parameters</returns>
    public IList<ParameterView> GetParameters()
    {
        // Parameters from config.json
        Dictionary<string, ParameterView> result = new(StringComparer.OrdinalIgnoreCase);
        foreach (var p in this._promptConfig.Input.Parameters)
        {
            result[p.Name] = new ParameterView(p.Name, p.Description, p.DefaultValue);
        }

        return result.Values.ToList();
    }

    /// <summary>
    /// Render the template using the information in the context
    /// </summary>
    /// <param name="executionContext">Kernel execution context helpers</param>
    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests. The default is <see cref="CancellationToken.None"/>.</param>
    /// <returns>Prompt rendered to string</returns>
    public async Task<string> RenderAsync(SKContext executionContext, CancellationToken cancellationToken)
    {
        return await this._templateEngine.RenderAsync(this._template, executionContext, cancellationToken).ConfigureAwait(false);
    }
}

File: typescript/src/semanticFunctions/promptTemplate.ts
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

import { CaseInsensitiveSet } from '../utils';
import { ILogger, NullLogger } from '../utils/logger';
import { SKContext } from '../orchestration';
import { ParameterView } from '../registry';
import { IPromptTemplateEngine } from '../templateEngine';
import { BlockTypes, VarBlock } from '../templateEngine/Blocks';
import { IPromptTemplate } from './iPromptTemplate';
import { IPromptTemplateConfig } from './promptTemplateConfig';

export class PromptTemplate implements IPromptTemplate {
    private readonly _template: string;
    private readonly _templateEngine: IPromptTemplateEngine;
    private readonly _log: ILogger = new NullLogger();
    private readonly _promptConfig: IPromptTemplateConfig;

    constructor(template: string, promptTemplateConfig: IPromptTemplateConfig, kernel: IKernel) {
        this._template = template;
        this._promptConfig = promptTemplateConfig;
        this._templateEngine = kernel.PromptTemplateEngine;
    }

    public getParameters(): ParameterView[] {
        const seen = new CaseInsensitiveSet<string>();
        const result: ParameterView[] = [];

        // Parameters from config.json
        this._promptConfig.input?.parameters?.forEach((p) => {
            if (p) {
                const view = new ParameterView(p.name, p.description, p.defaultValue);
                result.push(view);
                seen.add(p.name);
            }
        });

        // Parameters from the template
        this._templateEngine
            .extractBlocks(this._template)
            .filter((x) => x.type === BlockTypes.Variable)
            .map((x) => x as VarBlock)
            .filter((x) => !seen.has(x.name))
            .forEach((x) => result.push(new ParameterView(x.name)));

        return result;
    }

    public async render(executionContext: SKContext): Promise<string> {
        return await this._templateEngine.render(this._template, executionContext);
    }
}
